server-spi-private/src/main/java/org/keycloak/validate/builtin/EmailValidator.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.validate.builtin;
+
+import org.keycloak.validate.CompactValidator;
+import org.keycloak.validate.ValidationContext;
+import org.keycloak.validate.ValidationError;
+import org.keycloak.validate.ValidatorConfig;
+
+import java.util.regex.Pattern;
+
+public class EmailValidator implements CompactValidator {
+
+    public static final String ID = "email";
+
+    public static final EmailValidator INSTANCE = new EmailValidator();
+
+    public static final String MESSAGE_INVALID_EMAIL = "error-invalid-email";
+
+    // Actually allow same emails like angular. See ValidationTest.testEmailValidation()
+    private static final Pattern EMAIL_PATTERN = Pattern.compile(
+            "[a-zA-Z0-9!#$%&'*+/=?^_`{|}~.-]+@[a-zA-Z0-9-]+(\\.[a-zA-Z0-9-]+)*");
+
+    private EmailValidator() {
+    }
+
+    @Override
+    public String getId() {
+        return ID;
+    }
+
+    @Override
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (!(input instanceof String)) {
+            context.addError(new ValidationError(ID, inputHint, MESSAGE_INVALID_EMAIL, input));
```

Comment:
Should we rather throw some exception?  It's not exactly a validation error, the user providing the input is not responsible for this.

Reply:
I'd rather not throw an exception by default in that case, but we could configure the ValidationContext to throw an exception when a ValidationError is reported.

Reply:
Ok. My original assumption was that the user cannot affect the input type and therefore throwing incompatible input type at the validator is a dev's fault and as such is not a validation error. This is, of course, wrong – a user can in fact define input type (e.g. JSON payload in REST endpoints). Let's return a validation error.

But maybe we could have some predefined error message common for all validators for this case (incompatible input type)?